dotnet/src/Connectors/Connectors.Memory.Kusto/KustoMemoryStore.cs

```diff
@@ -0,0 +1,411 @@
+    {
+        return JsonConvert.DeserializeObject<MemoryRecordMetadata>(metadata)!;
+    }
+}
+
+internal static class KustoExtensions
+{
+    public static async ValueTask<T?> FirstOrDefaultAsync<T>(this IAsyncEnumerable<T> source, CancellationToken cancellationToken = default)
+    {
+        if (source == null)
+        {
+            throw new ArgumentNullException(nameof(source));
+        }
+
+        await foreach (var item in source.WithCancellation(cancellationToken).ConfigureAwait(false))
+        {
+            return item;
+        }
+
+        return default;
+    }
+}
+
+sealed internal class KustoMemoryRecord
+{
+    public string Key { get; set; }
+    public MemoryRecordMetadata Metadata { get; set; }
+    public Embedding<float> Embedding { get; set; }
+    public DateTime? Timestamp { get; set; }
+
+    public KustoMemoryRecord(MemoryRecord record) : this(record.Key, record.Metadata, record.Embedding, record.Timestamp?.UtcDateTime) { }
+
+    public KustoMemoryRecord(string key, MemoryRecordMetadata metadata, Embedding<float> embedding, DateTime? timestamp = null)
+    {
+        this.Key = key;
+        this.Metadata = metadata;
+        this.Embedding = embedding;
+        this.Timestamp = timestamp;
+    }
+
+    public KustoMemoryRecord(string key, string metadata, string? embedding, DateTime? timestamp = null)
+    {
+        this.Key = key;
+        this.Metadata = KustoSerializer.DeserializeMetadata(metadata);
+        this.Embedding = KustoSerializer.DeserializeEmbedding(embedding);
+        this.Timestamp = timestamp;
+    }
+
+    public MemoryRecord ToMemoryRecord()
+    {
+        return new MemoryRecord(this.Metadata, this.Embedding, this.Key, this.Timestamp);
+    }
+
+    public void WriteToCsvStream(CsvWriter streamWriter)
+    {
+        var jsonifiedMetadata = KustoSerializer.SerializeMetadata(this.Metadata);
+        var jsonifiedEmbedding = KustoSerializer.SerializeEmbedding(this.Embedding);
+        var isoFormattedDate = this.Timestamp?.ToString("o", CultureInfo.InvariantCulture) ?? string.Empty;
+
+        streamWriter.WriteField(this.Key);
+        streamWriter.WriteField(jsonifiedMetadata);
+        streamWriter.WriteField(jsonifiedEmbedding);
+        streamWriter.WriteField(isoFormattedDate);
+        streamWriter.CompleteRecord();
+    }
+}
+
+/// <summary>
+/// An implementation of <see cref="IMemoryStore"/> backed by a Kusto database.
+/// </summary>
+/// <remarks>The embedded data is saved to the Kusto database specified in the constructor.
+/// Similarity search capability is provided through a cosine similarity function (added on first collection creation). Use Kusto's "Table" to implement "Collection".
+/// </remarks>
+public class KustoMemoryStore : IMemoryStore
+{
+    private string _database;
+    private static ClientRequestProperties GetClientRequestProperties() => new()
+    {
+        Application = "SemanticKernel",
```

Comment:
I'm not sure I understand why it matters what is the user agent we report for our internal telemetry, but sure, I don't mind using a const. However, if we will build dashboards based on that, and it will change, that might break us.

Reply:
> I'm not sure I understand why it matters what is the user agent we report for our internal telemetry

The value itself doesn't matter, you are right. It's just the approach we are using to organize our code. It's also good to keep Application identifier consistent across different integrations.

> if we will build dashboards based on that, and it will change, that might break us

Value `Semantic-Kernel` most probably won't be changed.